lib/pkcs11h-openssl-1.1.1.c

```diff
@@ -0,0 +1,1761 @@
+}
+
+#ifndef OPENSSL_NO_RSA
+
+PKCS11H_BOOL
+__pkcs11h_md2ckm (
+	IN const EVP_MD * const md,
+	OUT CK_MECHANISM_TYPE * const hash_alg
+)
+{
+	_PKCS11H_ASSERT(md!=NULL);
+	_PKCS11H_ASSERT(hash_alg!=NULL);
+
+	switch (EVP_MD_type(md)) {
+		case NID_sha1:
+			*hash_alg = CKM_SHA_1;
+		break;
+		case NID_sha224:
+			*hash_alg = CKM_SHA224;
+		break;
+		case NID_sha256:
+			*hash_alg = CKM_SHA256;
+		break;
+		case NID_sha384:
+			*hash_alg = CKM_SHA384;
+		break;
+		case NID_sha512:
+			*hash_alg = CKM_SHA512;
+		break;
+		default:
+			_PKCS11H_LOG (PKCS11H_LOG_WARN, "PKCS#11: Unknown digest type (%d)", EVP_MD_type(md));
+			return FALSE;
+		break;
+	}
+	return TRUE;
+}
+
+static CK_RSA_PKCS_MGF_TYPE
+__pkcs11h_md2mgf (
+	IN const EVP_MD * const md,
+	OUT CK_RSA_PKCS_MGF_TYPE * const mgf
+)
+{
+	_PKCS11H_ASSERT(md!=NULL);
+	_PKCS11H_ASSERT(mgf!=NULL);
+
+	switch (EVP_MD_type(md)) {
+		case NID_sha1:
+			*mgf = CKG_MGF1_SHA1;
+		break;
+		case NID_sha224:
+			*mgf = CKG_MGF1_SHA224;
+		break;
+		case NID_sha256:
+			*mgf = CKG_MGF1_SHA256;
+		break;
+		case NID_sha384:
+			*mgf = CKG_MGF1_SHA384;
+		break;
+		case NID_sha512:
+			*mgf = CKG_MGF1_SHA512;
+		break;
+		default:
+			_PKCS11H_LOG (PKCS11H_LOG_WARN, "PKCS#11: Unknown mgf1 digest type (%d)", EVP_MD_type(md));
+			return FALSE;
+		break;
+	}
+	return TRUE;
+}
+
+static PKCS11H_BOOL
+__pkcs11h_get_ossl_pss_params (
+	IN EVP_PKEY_CTX *ctx,
+	OUT EVP_MD **md,
+	OUT EVP_MD **mgf1_md,
+	OUT int * const saltlen
+)
+{
+	_PKCS11H_ASSERT (ctx!=NULL);
+	_PKCS11H_ASSERT (md!=NULL);
+	_PKCS11H_ASSERT (mgf1_md!=NULL);
+	_PKCS11H_ASSERT (saltlen!=NULL);
+
+	if (
+		(EVP_PKEY_CTX_get_signature_md (ctx, md) <= 0) ||
+		(EVP_PKEY_CTX_get_rsa_mgf1_md (ctx, mgf1_md) <= 0) ||
+		(EVP_PKEY_CTX_get_rsa_pss_saltlen (ctx, saltlen) <= 0) ||
+		(*md == NULL) || (*mgf1_md == NULL)
+	) {
+		_PKCS11H_LOG (PKCS11H_LOG_WARN, "PKCS#11: failed to get pss params from ctx");
+		return FALSE;
+	}
+	_PKCS11H_DEBUG (
+		PKCS11H_LOG_DEBUG1,
+		"PKCS#11: __pkcs11h_get_pss_params: saltlen=%d md=%s mgf1_md=%s",
+		*saltlen, EVP_MD_name (*md), EVP_MD_name (*mgf1_md)
+	);
+	return TRUE;
+}
+
+static PKCS11H_BOOL
+__pkcs11h_get_pss_params (
+	IN EVP_PKEY_CTX *ctx,
+	OUT CK_RSA_PKCS_PSS_PARAMS *params
+) {
+
+	EVP_MD *md, *mgf1_md;
+	int saltlen;
+	EVP_PKEY *pkey;
+
+	_PKCS11H_ASSERT (ctx!=NULL);
+	_PKCS11H_ASSERT (params!=NULL);
+
+	pkey = EVP_PKEY_CTX_get0_pkey (ctx);
+	if (pkey==NULL)
+		return FALSE;
+
+	if (!__pkcs11h_get_ossl_pss_params (ctx, &md, &mgf1_md, &saltlen))
+		return FALSE;
+
+	if (saltlen == RSA_PSS_SALTLEN_DIGEST) {
+		saltlen = EVP_MD_size (md);
+	}
+	else if ((saltlen == RSA_PSS_SALTLEN_MAX) || (saltlen == RSA_PSS_SALTLEN_MAX_SIGN)) {
+		saltlen = EVP_PKEY_size (pkey) - EVP_MD_size (md) - 2;
+	}
+	if (((EVP_PKEY_bits (pkey) - 1) & 0x7) == 0) {
+		saltlen -= 1;
+	}
+	if (saltlen < 0) {
+		_PKCS11H_LOG (PKCS11H_LOG_WARN, "PKCS#11: pss invalid saltlen < 0 (%d)", saltlen);
+		return FALSE;
+	}
```

Comment:
are you sure openssl does not support a magic method flags set to perform this logic by itself?

Reply:
Not that I'm aware of. Looking at OpenSSL code, its default pkey_rsa_sign does the same check. That said, I have never seen it call sign() with anything but saltlen = -1, meaning use the digest length.